ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/DbChangelog/Unit.hs

```diff
@@ -0,0 +1,339 @@
+        ]
+      , testProperty "rolling back" $ withMaxSuccess samples $ conjoin
+        [ counterexample "rollback after extension is noop"
+          prop_rollbackAfterExtendIsNoop
+        , counterexample "prefixing back to anchor is rolling back volatile states"
+          prop_prefixBackToAnchorIsRollingBackVolatileStates
+        , counterexample "prefix back to volatile tip is a noop"
+          prop_rollBackToVolatileTipIsNoop
+        ]
+      , testProperty "extending adds head to volatile states"
+        $ withMaxSuccess samples prop_extendingAdvancesTipOfVolatileStates
+      , testProperty "pruning leaves at most maxRollback volatile states"
+        $ withMaxSuccess samples prop_pruningLeavesAtMostMaxRollbacksVolatileStates
+      ]
+
+
+
+{-------------------------------------------------------------------------------
+  Test setup
+-------------------------------------------------------------------------------}
+
+data TestLedger (mk :: MapKind) = TestLedger {
+  tlUtxos :: mk Key Int,
+  tlTip   :: Point TestLedger
+}
+
+nextState :: DbChangelog TestLedger -> TestLedger DiffMK
+nextState dblog = TestLedger {
+              tlTip = pointAtSlot $ nextSlot (getTipSlot old)
+            , tlUtxos = DiffMK mempty
+            }
+  where
+    old = DbChangelog.current $ anchorlessChangelog dblog
+    nextSlot = At . withOrigin 1 (+1)
+
+
+deriving instance Show (mk Key Int) => Show (TestLedger mk)
+
+instance GetTip TestLedger where
+  getTip = castPoint . tlTip
+
+data H = H deriving (Eq, Ord, Show, Generic)
+deriving anyclass instance NoThunks H
+type instance HeaderHash TestLedger = H
+
+instance StandardHash TestLedger
+
+deriving instance Eq (TestLedger EmptyMK)
+
+type instance Ledger.Key   TestLedger = Key
+type instance Ledger.Value TestLedger = Int
+
+instance HasLedgerTables TestLedger where
+  projectLedgerTables                     = LedgerTables . tlUtxos
+  withLedgerTables st    (LedgerTables x) = st { tlUtxos = x }
+
+data DbChangelogTestSetup = DbChangelogTestSetup {
+  -- The operations are applied on the right, i.e., the newest operation is at the head of the list.
+    operations          :: [Operation TestLedger]
+  , dbChangelogStartsAt :: WithOrigin SlotNo
+  }
+
+data Operation l = Extend (l DiffMK) | Prune SecurityParam
+deriving instance Show (l DiffMK) => Show (Operation l)
+
+data DbChangelogTestSetupWithRollbacks = DbChangelogTestSetupWithRollbacks
+  { testSetup :: DbChangelogTestSetup
+  , rollbacks :: Int
+  } deriving (Show)
+
+instance Show DbChangelogTestSetup where
+  show = ppShow . operations
+
+instance Arbitrary DbChangelogTestSetup where
+  arbitrary = sized $ \n -> do
+    slotNo <- oneof [pure Origin, At . SlotNo <$> chooseEnum (1, 1000)]
+    ops <- genOperations slotNo n
+    pure $ DbChangelogTestSetup
+      { operations = ops
+      , dbChangelogStartsAt = slotNo
+      }
+
+  -- TODO: Shrinking might not be optimal. Shrinking finds the shortest prefix of the list of
```

Comment:
reifying a TODO

Reply:
I think this is perfectly fine. We are indeed shrinking the commands to a minimal sequence.